### ISSUE TITLE
PyTorch version training issues

### ISSUE THREAD

**Issue by longlong161** [albrateanu/LYT-Net#14]
After training the PyTorch version for a while, the metrics will not change. I trained twice and this problem started at 454 epochs.
![image](https://github.com/user-attachments/assets/2df1b5d0-70bb-469f-aa0a-6a2381213e2b)
![image](https://github.com/user-attachments/assets/c42a743c-5176-405d-a815-0a0eb3c35f04)

By the way, can we also add SSIM evaluation metrics to the PyTorch version? Thank you

**Comment by albrateanu** [comment]
Hello. Will add SSIM evaluation today. 
About the training issues. I still need to fix the model forward pass a bit. Moreover, training on PyTorch has different behaviour to TensorFlow apparently, so I will still be checking to see how to bring it as close as possible to the TensorFlow initial implementation.

**Comment by longlong161** [comment]
> Hello. Will add SSIM evaluation today. About the training issues. I still need to fix the model forward pass a bit. Moreover, training on PyTorch has different behaviour to TensorFlow apparently, so I will still be checking to see how to bring it as close as possible to the TensorFlow initial implementation.

Okay, thank you for your reply. Do you still need to add files related to testing in the future? Looking forward to the complete PyTorch version

**Comment by stuhao251** [comment]
感觉是在那个轮次训练已经达到饱和了，哈哈哈

**Comment by albrateanu** [comment]
![image](https://github.com/user-attachments/assets/7b939367-bec1-4b62-93ab-362c30fcecb3)
@longlong161 Don't know exactly what the issue with training is for you. I didn't have that. Looks like a gradient problem. Please retry with the last code and let me know if it still persists.

@stuhao251 It might be the case, unfortunately. I still get better metrics when training with TF for some reason. However, the baseline PyTorch implementation of the model should be exactly the same as the TF version, as both the TF and Torch versions of the model have 44923 parameters. So please feel free to develop on it if you think there's space for improvement.

**Comment by longlong161** [comment]
@albrateanu I used the latest code you uploaded and encountered a new error that appeared on both machines.
![image](https://github.com/user-attachments/assets/ad2c0d41-9aa9-445a-bc35-c0731acb9735)
![image](https://github.com/user-attachments/assets/0e6c1d21-3b60-424e-9587-cffed0c559e0)

**Comment by GZY2000** [comment]
> @albrateanu I used the latest code you uploaded and encountered a new error that appeared on both machines. ![image](https://private-user-images.githubusercontent.com/52652419/352491895-ad2c0d41-9aa9-445a-bc35-c0731acb9735.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzODkyNTQsIm5iZiI6MTcyMjM4ODk1NCwicGF0aCI6Ii81MjY1MjQxOS8zNTI0OTE4OTUtYWQyYzBkNDEtOWFhOS00NDVhLWJjMzUtYzA3MzFhY2I5NzM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMxVDAxMjIzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE3YmFmNmYwNjllNThlMGVlODMzODE4YTBiNzczNmVhMWZlMjQ2Y2IwOWExNGY4NTNiNTdlMjAzNzE0NDgzY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.azyzlrPK8zwQSFChUOMWRozNjYhq232trp__5HIcH8U) ![image](https://private-user-images.githubusercontent.com/52652419/352491951-0e6c1d21-3b60-424e-9587-cffed0c559e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzODkyNTQsIm5iZiI6MTcyMjM4ODk1NCwicGF0aCI6Ii81MjY1MjQxOS8zNTI0OTE5NTEtMGU2YzFkMjEtM2I2MC00MjRlLTk1ODctY2ZmZWQwYzU1OWUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMxVDAxMjIzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4NDgyMWYxNjQ5M2ZkNzYyYTc2YjIzYzViMTkzMmEzNTdlNDM2YjIwN2I5ODJhMjYwNzM1ZTI5ZTAwZWY4Y2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oIIHsPqPpz5bcQRL00V4UOy3aTQNyfUS5mt1jOFiLww)

Hello, has pytorch version come out yet? Why is my result so much worse than the results of tensorflow in the paper?

**Comment by albrateanu** [comment]
@longlong161 I'm not sure how I can help with that. Please make sure the environment is ok and running PyTorch 2. And perhaps just attempt retraining more? Or try setting either the Smooth L1 loss or the MS-SSIM loss weight to 0? It seems like a gradient explosion problem and it's not very clear what causes it given that I didn't have that.
@GZY2000 I keep saying this, I am not sure. I am able to obtain at least around 0.8-1dB more on TensorFlow than on PyTorch. But it's a bit out of my knowledge.

**Comment by albrateanu** [comment]
@GZY2000 I have just posted PyTorch weights. On LOLv2 Real, PyTorch implementations has 0.61dB more than TensorFlow.

**Comment by GZY2000** [comment]
> @GZY2000 I have just posted PyTorch weights. On LOLv2 Real, PyTorch implementations has 0.61dB more than TensorFlow.

That's true, but synthetic is more than 2db off on this dataset, which is the best result you can get after a lot of training!

**Comment by climberwu** [comment]
Hello,May I ask which journal your paper has been submitted to?
